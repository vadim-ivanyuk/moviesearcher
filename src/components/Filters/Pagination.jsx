import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { onChangeFilters } from "../../redux/filters/filters.actions";

export const Pagination = () => {
  const filters = useSelector((store) => store.filters);
  const dispatch = useDispatch();

  const handleChange = (page) => () => {
    dispatch(onChangeFilters("page", page));
  };

  return (
    <div className="pagination form-group col-12 d-flex justify-content-center">
      <button
        type="button"
        className="btn btn-light col-4"
        onClick={handleChange(+filters.page - 1)}
        disabled={filters.page === 1}
      >
        Назад
      </button>
      <div className="col-4 total-pages d-flex align-items-center">{`${filters.page} из ${filters.total_pages}`}</div>
      <button
        type="button"
        className="btn btn-light col-4"
        onClick={handleChange(+filters.page + 1)}
        disabled={Number(filters.page === filters.total_pages)}
      >
        Вперед
      </button>
    </div>
  );
};
