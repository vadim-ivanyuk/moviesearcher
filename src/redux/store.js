import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import rootReducer from "./rootReducer";
import { ON_CHANGE_FILTERS } from "./filters/filters.types";

let currentPage;

const onChangePageMiddleware = ({ dispatch, getState }) => (next) => (
  action
) => {
  currentPage = getState().filters.page;
  if (
    action.type === ON_CHANGE_FILTERS &&
    action.payload.name !== "page" &&
    currentPage !== 1
  ) {
    dispatch({
      type: ON_CHANGE_FILTERS,
      payload: {
        name: "page",
        value: 1,
      },
    });
  }

  return next(action);
};

const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(thunk, onChangePageMiddleware))
);

export default store;
