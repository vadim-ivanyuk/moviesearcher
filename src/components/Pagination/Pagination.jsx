import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { onChangeFilters } from "../../redux";

export const Pagination = () => {
  const filters = useSelector((store) => store.filters);
  const dispatch = useDispatch();

  const handleChange = (page) => () => {
    dispatch(onChangeFilters({ name: "page", value: page }));
  };

  return (
    <div className="pagination form-group col-12 d-flex justify-content-center p-0">
      <button
        type="button"
        className="btn btn-light col-4"
        onClick={handleChange(+filters.page - 1)}
        disabled={filters.page === 1}
      >
        Назад
      </button>
      <div className="col-4 total-pages d-flex align-items-center justify-content-center p-0">{`${filters.page} из ${filters.total_pages}`}</div>
      <button
        type="button"
        className="btn btn-light col-4"
        onClick={handleChange(+filters.page + 1)}
        disabled={Number(filters.page === filters.total_pages) || false}
      >
        Вперед
      </button>
    </div>
  );
};
