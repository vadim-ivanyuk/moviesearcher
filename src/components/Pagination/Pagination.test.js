import React from "react";
import { configure, mount } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import store from "../../redux/store";
import { Pagination } from "./Pagination";

configure({ adapter: new Adapter() });

jest.mock();

describe("Pagination", () => {
  const pagination = mount(
    <Provider store={store}>
      <Pagination />
    </Provider>
  );
  const buttonPrev = pagination.find("button.btn").first();
  const buttonNext = pagination.find("button.btn").last();

  it("first render snap", () => {
    expect(shallowToJson(pagination)).toMatchSnapshot();
  });

  it("button click", () => {
    expect(buttonPrev.props().disabled).toEqual(true);

    expect(pagination.props().store.getState().filters.page).toEqual(1);

    for (let i = 1; i < 10; i++) {
      buttonNext.simulate("click");

      expect(pagination.props().store.getState().filters.page).toEqual(i + 1);
    }

    console.log(pagination.props());
  });

  it("last render snap", () => {
    expect(shallowToJson(pagination)).toMatchSnapshot();
  });
});
