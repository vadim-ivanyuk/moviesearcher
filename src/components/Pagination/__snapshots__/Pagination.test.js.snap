// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination first render snap 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Pagination>
    <div
      className="pagination form-group col-12 d-flex justify-content-center p-0"
    >
      <button
        className="btn btn-light col-4"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Назад
      </button>
      <div
        className="col-4 total-pages d-flex align-items-center justify-content-center p-0"
      >
        1 из 500
      </div>
      <button
        className="btn btn-light col-4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Вперед
      </button>
    </div>
  </Pagination>
</Provider>
`;

exports[`Pagination last render snap 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Pagination>
    <div
      className="pagination form-group col-12 d-flex justify-content-center p-0"
    >
      <button
        className="btn btn-light col-4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Назад
      </button>
      <div
        className="col-4 total-pages d-flex align-items-center justify-content-center p-0"
      >
        10 из 500
      </div>
      <button
        className="btn btn-light col-4"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Вперед
      </button>
    </div>
  </Pagination>
</Provider>
`;
